osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -255,7 +257,76 @@ private EncryptedTypedSecret retrieveSecret(
     return Optional.of(secretsList.iterator().next());
   }
 
-  public synchronized void storeSecrets(
+  public synchronized void storeOrResetNewSecrets(
+      final String deviceIdentification, final List<TypedSecret> typedSecrets) {
+    final List<TypedSecret> typedNewSecretsToStore = new ArrayList<>();
+    final List<TypedSecret> typedNewSecretsToReset = new ArrayList<>();
+    for (final TypedSecret typedSecret : typedSecrets) {
+      final boolean thereAreNoKeysWithStatusNew =
+          this.thereAreNoKeysWithStatusNew(deviceIdentification, typedSecret.getSecretType());
+      if (thereAreNoKeysWithStatusNew) {
+        typedNewSecretsToStore.add(typedSecret);
+      } else {
+        typedNewSecretsToReset.add(typedSecret);
+      }
+    }
+    this.storeSecrets(deviceIdentification, typedNewSecretsToStore);
+    typedNewSecretsToReset.forEach(
+        ts -> this.resetNewSecret(deviceIdentification, ts.getSecretType()));
+  }
+
+  private boolean thereAreNoKeysWithStatusNew(
+      final String deviceIdentification, final SecretType secretType) {
+    final int numberOfNewSecrets =
+        this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW);
+    return numberOfNewSecrets == 0;
+  }
+
+  private TypedSecret resetNewSecret(
+      final String deviceIdentification, final SecretType secretType) {
+    // the most recent key of all NEW keys of the device and type gets a new creation date and is
+    // returned. All other NEW keys of the device and type are set to status EXPIRED.
```

Comment:
In the discussion with the architect your doubt has been noted. Nevertheless otherwise has been decided. Further discussion should not take place in this story's review but on another platform. I'm not taking stand on whether your doubt is funded or not.